vmware-event-router/internal/processor/openfaas/openfaas.go

```diff
@@ -283,18 +281,21 @@ func (p *Processor) PushMetrics(ctx context.Context, ms metrics.Receiver) {
 
 // Shutdown performs a clean shutdown of the OpenFaaS processor. It must not be
 // called more than once and only after all inflight event processing requests
-// have finished to avoid a panic.
+// have finished to avoid a panic. If the processor has already been stopped
+// ErrStopped is returned.
 func (p *Processor) Shutdown(_ context.Context) error {
-	if !p.isStopped() {
-		p.lock.Lock()
-		p.stopped = true
-		p.lock.Unlock()
+	p.Logger.Infof("attempting graceful shutdown")
+	if p.isStopped() {
+		return ErrStopped
 	}
 
+	p.lock.Lock()
+	p.stopped = true
+	p.lock.Unlock()
+
 	// free resources - if shutdown is called when they're still inflight processor
```

Comment:
Also use wg with timeout to avoid panic and return with error?

Reply:
Will be tackled in follow-up PR

Reply:
Are you rethinking to not "intentionally cause a panic by writing to a closed channel"?

Everything is already being shutdown, and they didn't shutdown in the proper sequence, I think it's ok the panic.

On the other hand, an orderly shutdown does seem nice. Wait for the invocations to finish, then shutdown.

Reply:
The new code (separate PR) will try to prevent a possible panic by waiting for `Process()` (and related goroutines) to finish. However, there will be a timeout which will be enforced leaving the potential (low chance though) of a panic during shutdown when the timeout is exceeded.